feat: refactor tables to reuse common fields

adklempner opened the pull request:
Depends on: https://github.com/status-im/telemetry/pull/44

- Adds a device type field for all metrics
- Creates a new table for fields that are common across metrics
- Removes fields from above in existing tables, adds a foreign key to above
- Default to SQL time for `createdAt`
- Updates types and put/process functions to work with new table and foreign key

igor-sirotin commented on telemetry/sql/000014_device_type.up.sql:
Same here, is it possible to introduce some `RecordID` to each record stored in separate tables, and then introduce a new table that will hold all common fields? And just `LEFT JOIN` the tables on read.

igor-sirotin replied:
Obviously, it's a big change, and not required for this PR.
Yet, if we keep adding more fields, would be nice to consider.

adklempner replied:
> Same here, is it possible to introduce some `RecordID` to each record stored in separate tables, and then introduce a new table that will hold all common fields? And just `LEFT JOIN` the tables on read.

Would this mean we have to write to two tables each time we receive data?

igor-sirotin replied:
Yes, but I don't think that's a problem. Is it?

adklempner replied:
Added in latest commit

igor-sirotin commented on pkg/types/types.go:
```diff
@@ -21,62 +21,80 @@ type TelemetryRequest struct {
 
 type ErrorSendingEnvelope struct {
-	CreatedAt    int64        `json:"createdAt"`
+	Id           int          `json:"id"`
```
```suggestion
	ID           int          `json:"id"`
```

igor-sirotin commented on telemetry/common_fields.go:
```diff
@@ -0,0 +1,32 @@
+	"github.com/status-im/telemetry/pkg/types"
+)
+
+func InsertCommonFields(db *sql.DB, data types.CommonFieldsAccessor) (int, error) {
```
Why not just pass pointer to `CommonFields` here? I think using interface here is a bit overcomplicating.

```suggestion
func InsertCommonFields(db *sql.DB, data *types.CommonFields) (int, error) {
```

adklempner replied:
Ah I didn't know it could work like this! much simpler, thanks

igor-sirotin commented on telemetry/sql/000014_device_type.up.sql:
```diff
@@ -0,0 +1,6 @@
+ALTER TABLE peercount ADD COLUMN deviceType VARCHAR(255);
+ALTER TABLE receivedMessages ADD COLUMN deviceType VARCHAR(255);
+ALTER TABLE receivedEnvelopes ADD COLUMN deviceType VARCHAR(255);
+ALTER TABLE sentEnvelopes ADD COLUMN deviceType VARCHAR(255);
+ALTER TABLE errorSendingEnvelope ADD COLUMN deviceType VARCHAR(255);
+ALTER TABLE peerConnFailure ADD COLUMN deviceType VARCHAR(255);
```
Do we still need to add `deviceType` to all tables?

adklempner replied:
Keeping it in as it's included in separate PR that will be merged first https://github.com/status-im/telemetry/pull/44

igor-sirotin commented on telemetry/sql/000015_common_fields.up.sql:
```diff
@@ -0,0 +1,118 @@
+BEGIN
+
+    -- Add the new column for the foreign key
+    EXECUTE format('ALTER TABLE %I ADD COLUMN IF NOT EXISTS commonFieldsId INTEGER', table_name);
```
Not pushing on termins here, just sharing opniong.

I would call this column just `id`. And the `commonFields`, would simply be `TelemetryRecord`. And then there're difference extensions of these records that reference the main record by ID.

```suggestion
    EXECUTE format('ALTER TABLE %I ADD COLUMN IF NOT EXISTS id INTEGER', table_name);
```

adklempner replied:
Do you mean multiple rows in other tables sharing the same `TelemetryRecord`? e.g. a row in `receivedMessages` and a row in `peercount` with the same value in column `id`?

I think this would work but would require removing the timestamp from `TelemetryRecord` and putting it back into the other tables.

igor-sirotin replied:
Nah, never mind. After second thought on this, I decided to google how to do it right 😄 
What we're doing here is called [Class Table Inheritance](https://stackoverflow.com/a/3579462/7449458).

igor-sirotin replied:
So my suggestion would be just to rename `CommonFields` to `TelemetryRecord` (or simply `Record`).
And consequently rename `commonFieldsId` -> `record_id`. 

But this is not just my subjective opinion, leaving this at your decision ofc.

adklempner replied:
Renamed in latest commit

igor-sirotin commented on telemetry/sql/000015_common_fields.up.sql:
```diff
@@ -0,0 +1,118 @@
+            RAISE NOTICE 'Foreign key constraint already exists on %', table_name;
+    END;
+
+    -- Remove columns that are now in commonFields
```
Any chance to migrate current telemetry to the new table as well?

igor-sirotin commented on telemetry/receivedmessage.go:
```diff
@@ -99,8 +98,11 @@ func queryReceivedMessagesBetween(db *sql.DB, startsAt time.Time, endsAt time.Ti
+	err := db.QueryRow(`
+		SELECT COUNT(*) 
+		FROM receivedMessages rm
+		JOIN commonFields cf ON rm.id = cf.id
```
Looks like it should be `rm.commonFieldsId`

igor-sirotin commented on telemetry/receivedmessage.go:
```diff
@@ -110,8 +112,11 @@ func didReceivedMessageBeforeAndAfterInChat(db *sql.DB, receiverPublicKey string
+	err = db.QueryRow(`
+		SELECT COUNT(*)
+		FROM receivedMessages rm
+		JOIN commonFields cf ON rm.id = cf.id
```
Looks like it should be `rm.commonFieldsId`

igor-sirotin commented on telemetry/receivedmessage.go:
```diff
@@ -65,7 +67,7 @@ func (r *ReceivedMessage) put(db *sql.DB) error {
 
 func queryReceivedMessagesBetween(db *sql.DB, startsAt time.Time, endsAt time.Time) ([]*types.ReceivedMessage, error) {
-	rows, err := db.Query(fmt.Sprintf("SELECT id, chatId, messageHash, messageId, receiverKeyUID, peerId, nodeName, sentAt, topic, messageType, messageSize, createdAt, pubSubTopic FROM receivedMessages WHERE sentAt BETWEEN %d and %d", startsAt.Unix(), endsAt.Unix()))
+	rows, err := db.Query(fmt.Sprintf("SELECT id, chatId, messageHash, messageId, receiverKeyUID, sentAt, topic, messageType, messageSize, pubSubTopic FROM receivedMessages WHERE sentAt BETWEEN %d and %d", startsAt.Unix(), endsAt.Unix()))
```
Shouldn't we `LEFT JOIN commonFields` here? 🤔

igor-sirotin reviewed (commented):
Thanks for the improvements!
I don't have any strong objections, mostly suggestions.
Also, I would recommend to keep this PR only about adding `deviceType`, and open a separate PR for the `commonFields`. It would help to to block one change with the other one. And the review would be simpler for the folks :)

igor-sirotin commented:
@richard-ramos please take a look as well

richard-ramos commented on telemetry/peer_count.go:
```diff
@@ -56,24 +58,24 @@ func (r *PeerConnFailure) process(db *sql.DB, errs *MetricErrors, data *types.Te
 	}
 
-	stmt, err := db.Prepare("INSERT INTO peerConnFailure (timestamp, nodeName, nodeKeyUid, peerId, failedPeerId, failureCount, statusVersion, createdAt) VALUES ($1, $2, $3, $4, $5, $6, $7, $8) RETURNING id;")
+	commonFieldsId, err := InsertCommonFields(db, r.data)
```
See comment about using a trx

richard-ramos commented on telemetry/receivedenvelope.go:
```diff
@@ -15,10 +14,15 @@ type ReceivedEnvelope struct {
-							topic, receiverKeyUID, nodeName, processingError, statusVersion)
-							VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9)
+
+	commonFieldsId, err := InsertCommonFields(db, r.data)
```
See comment about using a trx

richard-ramos commented on telemetry/receivedenvelope.go:
```diff
@@ -145,10 +148,14 @@ func (e *ErrorSendingEnvelope) process(db *sql.DB, errs *MetricErrors, data *typ
-	stmt, err := db.Prepare(`INSERT INTO errorSendingEnvelope (messageHash, sentAt, createdAt, pubsubTopic,
-		topic, senderKeyUID, peerId, nodeName, publishMethod, statusVersion, error)
-		VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11)
+	commonFieldsId, err := InsertCommonFields(db, e.data.SentEnvelope)
```
See comment about using a trx

richard-ramos commented on telemetry/receivedmessage.go:
```diff
@@ -35,27 +35,29 @@ func (r *ReceivedMessage) process(db *sql.DB, errs *MetricErrors, data *types.Te
 
 func (r *ReceivedMessage) put(db *sql.DB) error {
-	stmt, err := db.Prepare("INSERT INTO receivedMessages (chatId, messageHash, messageId, receiverKeyUID, peerId, nodeName, sentAt, topic, messageType, messageSize, createdAt, pubSubTopic, statusVersion) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13) RETURNING id;")
+	commonFieldsId, err := InsertCommonFields(db, r.data)
```
See comment about using a transaction

adklempner commented on pkg/types/types.go:
```diff
@@ -21,62 +21,80 @@ type TelemetryRequest struct {
-	CreatedAt       int64  `json:"createdAt"`
 	PubsubTopic     string `json:"pubsubTopic"`
 	Topic           string `json:"topic"`
 	SenderKeyUID    string `json:"senderKeyUID"`
```
status-go uses UID for all json so I set it to that

igor-sirotin replied:
I guess now it's too late to change it anyway, it would be a breaking change

adklempner commented on telemetry/aggregator_test.go:
```diff
@@ -28,52 +29,28 @@ func NewMock() *sql.DB {
-		log.Fatalf("an error '%s' was not expected when dropping the table", err)
-	}
+	for _, table := range tables {
+		_, err := db.Exec(fmt.Sprintf("DROP TABLE IF EXISTS %s", table))
```
fun fact, turns out it only works with:
```
db.Exec(`DROP TABLE IF EXISTS $1`, table)
```

richard-ramos replied:
Damn, that's annoying :(

igor-sirotin commented on telemetry/peer_count.go:
```diff
@@ -15,34 +15,47 @@ type PeerCount struct {
 	}
 
-	stmt, err := db.Prepare("INSERT INTO peerCount (timestamp, nodeName, nodeKeyUid, peerId, peerCount, statusVersion, createdAt) VALUES ($1, $2, $3, $4, $5, $6, $7) RETURNING id;")
+	tx, err := db.BeginTx(context.Background(), nil)
```
Not a big deal, but it's better to actually pass some real context

richard-ramos commented on lib/common/test_utils.go:
```diff
@@ -111,6 +89,11 @@ func DropTables(db *sql.DB) {
 		log.Fatalf("an error '%s' was not expected when dropping the table", err)
 	}
 
+	_, err = tx.Exec("DROP TABLE IF EXISTS telemetryRecord")
```
Should this table be included in the list on line 31?

adklempner replied:
It needs to be dropped after all other tables and indexes

richard-ramos commented on lib/metrics/common_fields.go:
```diff
@@ -0,0 +1,31 @@
+)
+
+func InsertTelemetryRecord(tx *sql.Tx, data *types.TelemetryRecord) (int, error) {
+	stmt, err := tx.Prepare(`
+		INSERT INTO telemetryRecord (nodeName, peerId, statusVersion, deviceType)
+		VALUES ($1, $2, $3, $4)
+		RETURNING id;
+	`)
```
As described in https://github.com/status-im/telemetry/issues/43 consider using exec, or if this insertion is going to be called frequently, maybe it's a good idea to extract the prepared statement and make it available during the entire duration of the program.

richard-ramos commented on lib/metrics/peer_count.go:
```diff
@@ -16,35 +16,47 @@ type PeerCount struct {
 	}
 
-	stmt, err := db.Prepare("INSERT INTO peerCount (timestamp, nodeName, nodeKeyUid, peerId, peerCount, statusVersion, createdAt, deviceType) VALUES ($1, $2, $3, $4, $5, $6, $7, $8) RETURNING id;")
+	tx, err := db.BeginTx(context.Background(), nil)
```
Ideally the context should be passed down. Maybe this one could be used https://github.com/status-im/telemetry/blob/master/telemetry/server.go#L39 ?

richard-ramos commented on lib/metrics/peer_count.go:
```diff
@@ -16,35 +16,47 @@ type PeerCount struct {
-		r.StatusVersion,
-		r.CreatedAt,
-		r.DeviceType,
+	peerCountStmt, err := tx.Prepare(`
+		INSERT INTO peerCount (common_fields_id, nodeKeyUid, peerCount)
+		VALUES ($1, $2, $3)
+		RETURNING id;
+	`)
```
Consider using exec vs prepare https://github.com/status-im/telemetry/issues/43

richard-ramos commented on lib/metrics/peer_count.go:
```diff
@@ -58,36 +70,46 @@ type PeerConnFailure struct {
+		return err
+	}
+
+	tx, err := db.BeginTx(context.Background(), nil)
```
Ideally the context should be passed down. Maybe this one could be used https://github.com/status-im/telemetry/blob/master/telemetry/server.go#L39 ?

richard-ramos commented on lib/metrics/receivedenvelope.go:
```diff
@@ -16,10 +16,20 @@ type ReceivedEnvelope struct {
-	stmt, err := db.Prepare(`INSERT INTO receivedEnvelopes (messageHash, sentAt, createdAt, pubsubTopic,
-							topic, receiverKeyUID, nodeName, processingError, statusVersion, deviceType)
-							VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10)
+	tx, err := db.BeginTx(context.Background(), nil)
```
Ideally the context should be passed down. Maybe this one could be used https://github.com/status-im/telemetry/blob/master/telemetry/server.go#L39 ?

richard-ramos commented on lib/metrics/peer_count.go:
```diff
@@ -58,36 +70,46 @@ type PeerConnFailure struct {
 	}
 
-	stmt, err := db.Prepare("INSERT INTO peerConnFailure (timestamp, nodeName, nodeKeyUid, peerId, failedPeerId, failureCount, statusVersion, createdAt, deviceType) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9) RETURNING id;")
+	stmt, err := db.Prepare("INSERT INTO peerConnFailure (common_fields_id, nodeKeyUid, failedPeerId, failureCount) VALUES ($1, $2, $3, $4) RETURNING id;")
```
See prev comment about using exec.

richard-ramos commented on lib/metrics/receivedenvelope.go:
```diff
@@ -16,10 +16,20 @@ type ReceivedEnvelope struct {
+		return err
+	}
+
+	stmt, err := tx.Prepare(`INSERT INTO receivedEnvelopes (recordId, messageHash, sentAt, pubsubTopic,
```
See prev comment about using exec.

richard-ramos commented on lib/metrics/receivedenvelope.go:
```diff
@@ -92,10 +101,20 @@ type SentEnvelope struct {
+		return err
+	}
+
+	stmt, err := tx.Prepare(`INSERT INTO sentEnvelopes (recordId, messageHash, sentAt, pubsubTopic,
```
See prev comment about using exec.

richard-ramos commented on lib/metrics/receivedenvelope.go:
```diff
@@ -152,14 +169,24 @@ type ErrorSendingEnvelope struct {
-	stmt, err := db.Prepare(`INSERT INTO errorSendingEnvelope (messageHash, sentAt, createdAt, pubsubTopic,
-		topic, senderKeyUID, peerId, nodeName, publishMethod, statusVersion, error, deviceType)
-		VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12)
+	tx, err := db.BeginTx(context.Background(), nil)
```
See comment about passing down context

richard-ramos commented on lib/metrics/receivedmessage.go:
```diff
@@ -40,39 +41,53 @@ func (r *ReceivedMessage) Clean(db *sql.DB, before int64) (int64, error) {
 
 func (r *ReceivedMessage) Put(db *sql.DB) error {
-	stmt, err := db.Prepare("INSERT INTO receivedMessages (chatId, messageHash, messageId, receiverKeyUID, peerId, nodeName, sentAt, topic, messageType, messageSize, createdAt, pubSubTopic, statusVersion, deviceType) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14) RETURNING id;")
+	tx, err := db.BeginTx(context.Background(), nil)
```
See prev comment about passing down the context

richard-ramos reviewed (approved):
LGTM. Just left few nitpicks about things that probably werent introduced in this PR and might be worked on later!

adklempner commented:
> LGTM. Just left few nitpicks about things that probably werent introduced in this PR and might be worked on later!

Replaced all `Prepare` with `QueryRow` (not `Exec` because Postgres driver doesn't support returning any values this way, and we want the ID) and updated to use passed down context. The latter was really easy to do thanks to the refactoring by @vpavlin !
